itemlist.lua

```diff
@@ -0,0 +1,231 @@
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+--]]
+
+--[[ The item list dialog can only be displayed from within a node formspec.
+To open the item list dialog, add a do_file :
+
+local show_item_list = dofile(
+	minetest.get_modpath(minetest.get_current_modname())..'/itemlist.lua')
+
+...
+
+-- From your on_receive_field, call :
+show_item_list(player, title, item_list, pos_context)
+
+player: player name or object
+title: title to be displayed at the top of the dialog
+item_list: a list of item names
+pos_context: position of the context node (node that openned the formspec)
+]]--
+
+-- Formspec style
+
+local style={
+	linesize = 0.9, -- Line height
+	colsize = 4,    -- Column width
+	lines = 8,      -- Number of lines of items displayed
+	cols = 3,       -- Number of columns of items displayed
+}
+
+local modname = minetest.get_current_modname()
+
+-- Context management functions (surely many improvements to do)
+
+local contexts = {}
+
+local function get_player_name(player)
+	if type(player) == 'string' then return player end
+	if type(player) == 'userdata' and player.get_player_name then
+		return player:get_player_name()
+	end
+	minetest.log('warning',	'['..modname..'] get_player_name could not identify player.')
+end
+
+minetest.register_on_leaveplayer(function(player)
+	local playername = get_player_name(player)
+	if playername then contexts[playername] = nil end
+end)
+
+local function new_context(player, context)
+	local playername = get_player_name(player)
+	if playername then
+		contexts[playername] = context
+		contexts[playername].playername = playername
+		return contexts[playername]
+    end
+end
+
+local function get_context(player)
+	local playername = get_player_name(player)
+	if playername then
+		if contexts[playername] then
+			return contexts[playername]
+		else
+			minetest.log('warning', '['..modname..'] Context not found for player "'..playername..'"')
+		end
+	end
+end
+
+local function update_context(player, changes)
+	local playername = get_player_name(player)
+	if playername then
+		if not contexts[playername] then
+			contexts[playername] = { playername = playername }
+		end
+		for key, value in pairs(changes) do
+			contexts[playername][key] = value
+		end
+	end
+end
+
+-- Show node formspec functions
+
+local function show_node_formspec(player, pos)
+	local meta = minetest.get_meta(pos)
+	local playername = get_player_name(player)
+
+	-- Decontextualize formspec
+	local fs = meta:get_string('formspec')
+
+	-- Change context and currrent_name references to nodemeta references
+	fs = fs:gsub("current_name", "nodemeta:"..pos.x..","..pos.y..","..pos.z)
```

Comment:
Looks hacky. Consider using a template to fill in the values using `string.format`

Reply:
Done in 1c132d84b0bd4a862c6d29f73a83d6c1ee372259

Reply:
Actually I meant the entire formspec but well, :man_shrugging: I guess that's OK too.

Reply:
Yes, it is hacky but unfortunately it seems that it is the only way to do that.